File: app/src/main/java/com/kobrakid/retroachievements/fragment/AchievementDistributionFragment.java
package com.kobrakid.retroachievements.fragment;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Build;
import android.os.Bundle;
import android.util.SparseIntArray;
import android.util.TypedValue;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.github.mikephil.charting.charts.LineChart;
import com.github.mikephil.charting.components.Description;
import com.github.mikephil.charting.components.XAxis;
import com.github.mikephil.charting.data.Entry;
import com.github.mikephil.charting.data.LineData;
import com.github.mikephil.charting.data.LineDataSet;
import com.github.mikephil.charting.highlight.Highlight;
import com.github.mikephil.charting.listener.OnChartValueSelectedListener;
import com.kobrakid.retroachievements.R;
import com.kobrakid.retroachievements.RAAPICallback;
import com.kobrakid.retroachievements.RAAPIConnection;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.select.Elements;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.SortedMap;
import java.util.TreeMap;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AchievementDistributionFragment extends Fragment implements RAAPICallback {

    private LineChart achievementDistro = null;
    private SortedMap<Integer, Integer> data;
    private boolean isActive = false, isAPIActive = false;

    public AchievementDistributionFragment() {
    }

    @SuppressLint("UseSparseArrays")
    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        setRetainInstance(true);

        final View view = inflater.inflate(R.layout.view_pager_achievement_distribution, container, false);

        achievementDistro = view.findViewById(R.id.game_details_achievement_distribution);
        achievementDistro.setOnChartValueSelectedListener(new OnChartValueSelectedListener() {
            @Override
            public void onValueSelected(Entry e, Highlight h) {
                if (isActive) {
                    ((TextView) view.findViewById(R.id.game_details_chart_hints))
                            .setText(getResources().getQuantityString(
                                    R.plurals.achievement_chart_hints,
                                    (int) e.getY(),
                                    (int) e.getY(),
                                    (int) e.getX()));
                }
            }

            @Override
            public void onNothingSelected() {
                if (isActive) {
                    ((TextView) view.findViewById(R.id.game_details_chart_hints)).setText("");
                }
            }
        });

        if (savedInstanceState == null) {
            data = new TreeMap<>();
            isAPIActive = true;
            new RAAPIConnection(getContext()).GetAchievementDistribution(Objects.requireNonNull(getArguments()).getString("GameID"), this);
        } else if (!isAPIActive) {
            populateChartData(view);
        }

        return view;
    }

    @Override
    public void onPause() {
        super.onPause();
        isActive = false;
    }

    @Override
    public void onResume() {
        super.onResume();
        isActive = true;
    }

    @Override
    public void onStart() {
        super.onStart();
        isActive = true;
    }

    @Override
    public void callback(int responseCode, String response) {
        if (!isActive)
            return;
        if (responseCode == RAAPIConnection.RESPONSE_GET_ACHIEVEMENT_DISTRIBUTION) {
            new AchievementDistributionChartAsyncTask(this, data).execute(response);
        }
        isAPIActive = false;
    }

    private void populateChartData(View view) {
        final Context context = getContext();
        if (context != null) {
            if (data.size() > 0) {
                // Set chart data
                List<Entry> entries = new ArrayList<>();
                for (Integer key : data.keySet()) {
                    entries.add(new Entry(key, data.get(key)));
                }
                LineDataSet dataSet = new LineDataSet(entries, "");
                dataSet.setDrawFilled(true);
                LineData lineData = new LineData(dataSet);
                lineData.setDrawValues(false);

                // Set chart colors
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                    TypedValue accentColor = new TypedValue(), primaryColor = new TypedValue();
                    context.getTheme().resolveAttribute(R.attr.colorAccent, accentColor, true);
                    context.getTheme().resolveAttribute(R.attr.colorPrimary, primaryColor, true);
                    achievementDistro.getAxisLeft().setTextColor(primaryColor.data);
                    achievementDistro.getXAxis().setTextColor(primaryColor.data);
                    dataSet.setCircleColor(accentColor.data);
                    dataSet.setColor(accentColor.data);
                    dataSet.setCircleHoleColor(accentColor.data);
                    dataSet.setFillColor(accentColor.data);
                }

                // Set chart axes
                achievementDistro.getAxisRight().setEnabled(false);
                achievementDistro.getLegend().setEnabled(false);
                achievementDistro.getXAxis().setPosition(XAxis.XAxisPosition.BOTTOM);
                achievementDistro.getAxisLeft().setAxisMinimum(0f);

                // Set chart description
                Description description = new Description();
                description.setText("");
                achievementDistro.setDescription(description);

                // Set chart finalized data
                achievementDistro.setData(lineData);

                // Redraw chart
                achievementDistro.invalidate();
            }
            view.findViewById(R.id.game_details_achievement_distro_loading).setVisibility(View.GONE);
            achievementDistro.setVisibility(View.VISIBLE);
        }
    }

    private static class AchievementDistributionChartAsyncTask extends AsyncTask<String, Integer, SortedMap<Integer, Integer>> {

        private final WeakReference<AchievementDistributionFragment> fragmentReference;
        private final WeakReference<SortedMap<Integer, Integer>> dataReference;

        AchievementDistributionChartAsyncTask(AchievementDistributionFragment fragment, SortedMap<Integer, Integer> data) {
            this.fragmentReference = new WeakReference<>(fragment);
            this.dataReference = new WeakReference<>(data);
        }

        @Override
        protected SortedMap<Integer, Integer> doInBackground(String... strings) {
            String response = strings[0];
            Document document = Jsoup.parse(response);
            Elements scripts = document.getElementsByTag("script");
            String rows = scripts.get(1).dataNodes().get(0).getWholeData();
            rows = rows.substring(rows.indexOf("dataTotalScore.addRows("));
            rows = rows.substring(0, rows.indexOf(");"));

            Pattern p1 = Pattern.compile("v:(\\d+),");
            Matcher m1 = p1.matcher(rows);
            Pattern p2 = Pattern.compile(",\\s(\\d+)\\s]");
            Matcher m2 = p2.matcher(rows);
            SparseIntArray achievementTotals = new SparseIntArray();
            while (m1.find() && m2.find()) {
                achievementTotals.put(Integer.parseInt(m1.group(1)), Integer.parseInt(m2.group(1)));
            }

            @SuppressLint("UseSparseArrays") SortedMap<Integer, Integer> chartData = new TreeMap<>();
            for (int i = 0; i < achievementTotals.size(); i++) {
                chartData.put(i + 1, achievementTotals.get(i + 1));
            }
            return chartData;
        }

        @Override
        protected void onPostExecute(SortedMap<Integer, Integer> chartData) {
            super.onPostExecute(chartData);
            final AchievementDistributionFragment fragment = fragmentReference.get();
            final SortedMap<Integer, Integer> data = dataReference.get();
            if (fragment != null && data != null) {
                data.putAll(chartData);
                fragment.populateChartData(fragment.getView());
            }
        }
    }
}

File: app/src/main/java/com/kobrakid/retroachievements/fragment/AchievementDistributionFragment.kt
package com.kobrakid.retroachievements.fragment

import android.annotation.SuppressLint
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.github.mikephil.charting.charts.LineChart
import com.github.mikephil.charting.components.Description
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.data.Entry
import com.github.mikephil.charting.data.LineData
import com.github.mikephil.charting.data.LineDataSet
import com.github.mikephil.charting.highlight.Highlight
import com.github.mikephil.charting.listener.OnChartValueSelectedListener
import com.kobrakid.retroachievements.R
import com.kobrakid.retroachievements.RetroAchievementsApi
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers.Default
import kotlinx.coroutines.Dispatchers.IO
import kotlinx.coroutines.Dispatchers.Main
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jsoup.Jsoup
import java.util.regex.Pattern

class AchievementDistributionFragment : Fragment() {

    private var achievementDistributionChart: LineChart? = null
    private var chartData = mutableListOf<Entry>()
    private var isActive = false

    @SuppressLint("UseSparseArrays")
    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        retainInstance = true
        val view = inflater.inflate(R.layout.view_pager_achievement_distribution, container, false)
        achievementDistributionChart = view.findViewById(R.id.game_details_achievement_distribution)
        achievementDistributionChart?.setOnChartValueSelectedListener(object : OnChartValueSelectedListener {
            override fun onValueSelected(e: Entry, h: Highlight) {
                if (isActive) {
                    view.findViewById<TextView>(R.id.game_details_chart_hints).text = resources.getQuantityString(
                            R.plurals.achievement_chart_hints,
                            e.y.toInt(),
                            e.y.toInt(),
                            e.x.toInt())
                }
            }

            override fun onNothingSelected() {
                if (isActive) {
                    view.findViewById<TextView>(R.id.game_details_chart_hints).text = ""
                }
            }
        })
        if (savedInstanceState == null && arguments != null) {
            val ctx = context?.applicationContext
            val id = arguments?.getString("GameID", "0")
            CoroutineScope(IO).launch {
                if (ctx != null && id != null)
                    RetroAchievementsApi.ScrapeGameInfoFromWeb(ctx, id) { parseAchievementDistribution(view, it) }
            }
        } else {
            populateChartData(view)
        }
        return view
    }

    override fun onPause() {
        super.onPause()
        isActive = false
    }

    override fun onResume() {
        super.onResume()
        isActive = true
    }

    override fun onStart() {
        super.onStart()
        isActive = true
    }

    private suspend fun parseAchievementDistribution(view: View, response: Pair<RetroAchievementsApi.RESPONSE, String>) {
        when (response.first) {
            RetroAchievementsApi.RESPONSE.ERROR -> {
                Log.w(TAG, response.second)
            }
            RetroAchievementsApi.RESPONSE.SCRAPE_GAME_PAGE -> {
                withContext(Default) {
                    Jsoup.parse(response.second)
                            .getElementsByTag("script")
                            .filter { it.html().startsWith("google.load('visualization'") }
                            .map {
                                val rows = it.dataNodes()[0].wholeData
                                val p1 = Pattern.compile("v:(\\d+),")
                                val m1 = p1.matcher(rows)
                                val p2 = Pattern.compile(",\\s(\\d+)\\s]")
                                val m2 = p2.matcher(rows)
                                while (m1.find() && m2.find()) {
                                    chartData.add(
                                            Entry(
                                                    m1.group(1)?.toFloat() ?: 0f,
                                                    m2.group(1)?.toFloat() ?: 0f
                                            )
                                    )
                                }
                            }
                }
                withContext(Main) {
                    populateChartData(view)
                }
            }
            else -> {
                Log.v(TAG, "${response.first}: ${response.second}")
            }
        }
    }

    private fun populateChartData(view: View) {
        val dataSet = LineDataSet(chartData, "")
        dataSet.setDrawFilled(true)
        val lineData = LineData(dataSet)
        lineData.setDrawValues(false)

        // Set chart colors
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            val accentColor = TypedValue()
            val primaryColor = TypedValue()
            context?.theme?.resolveAttribute(R.attr.colorAccent, accentColor, true)
            context?.theme?.resolveAttribute(R.attr.colorPrimary, primaryColor, true)
            achievementDistributionChart?.axisLeft?.textColor = primaryColor.data
            achievementDistributionChart?.xAxis?.textColor = primaryColor.data
            dataSet.setCircleColor(accentColor.data)
            dataSet.color = accentColor.data
            dataSet.circleHoleColor = accentColor.data
            dataSet.fillColor = accentColor.data
        }

        // Set chart axes
        achievementDistributionChart?.axisRight?.isEnabled = false
        achievementDistributionChart?.legend?.isEnabled = false
        achievementDistributionChart?.xAxis?.position = XAxis.XAxisPosition.BOTTOM
        achievementDistributionChart?.axisLeft?.axisMinimum = 0f

        // Set chart description
        val description = Description()
        description.text = ""
        achievementDistributionChart?.description = description

        // Set chart finalized data
        achievementDistributionChart?.data = lineData

        // Redraw chart
        achievementDistributionChart?.invalidate()

        view.findViewById<View>(R.id.game_details_achievement_distro_loading).visibility = View.GONE
        achievementDistributionChart?.visibility = View.VISIBLE
    }

    companion object {
        private val TAG = AchievementDistributionFragment::class.java.simpleName
    }
}